File: 10171.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> floydWarshall(vector<vector<pair<int,int>>> &G, map<pair<char,char>, char> &types, char type, map<int,char> &intToChar){
    int n = G.size(), i, j, k;
    vector<vector<int>> dist(n, vector<int>(n,1000));
    for(i = 0 ; i < n ; i++){
        for(j = 0 ; j < G[i].size() ; j++){
            if(types[{intToChar[i], intToChar[G[i][j].first]}] == type) dist[i][G[i][j].first] = G[i][j].second;
        }
        dist[i][i] = 0;
    }

    for(k = 0 ; k < n ; k++)
        for(i = 0 ; i < n ; i++)
            for(j = 0 ; j < n ; j++){
                char a = intToChar[i], b = intToChar[j], c = intToChar[k];
                if(types.find({a,c}) != types.end() and types.find({c,b}) != types.end() and types.find({a,b}) != types.end() and types[{a,c}]==type and types[{c,b}]==type and types[{a,b}]==type){
                    if(dist[i][k] + dist[k][j] < dist[i][j])
                        dist[i][j] = dist[i][k] + dist[k][j];
                    
                }
                    
            }
    return dist;
}

int main(){
    int i, edgesCnt, j, weight, n, young, elder, ans;
    while(true){
        char from, to;
        set<char> vertices;
        map<pair<char,char>, int> weights;
        map<pair<char,char>, char> types;
        cin >> edgesCnt;
        if(edgesCnt == 0) break;
        for(i = 0 ; i < edgesCnt ; i++){
            char type, dir, u, v;
            cin >> type >> dir >> u >> v; cin >> weight;
            vertices.insert(u); vertices.insert(v);
            weights[{u,v}] = weight;
            types[{u,v}] = type;
            types[{u,u}] = types[{v,v}] = 'x';
            if(dir == 'B'){
                weights[{v,u}] = weight;
                types[{v,u}] = type;
            }
        }
        n = vertices.size(); i = 0;
        map<char,int> charToInt; map<int,char> intToChar;
        set<char>::iterator it = vertices.begin();
        for(it ; it != vertices.end() ; it++){
            charToInt[*it] = i;
            intToChar[i++] = *it;
        }
        vector<vector<pair<int,int>>> G(n, vector<pair<int,int>>(0,{0,0}));
        map<pair<char,char>, int>::iterator it1 = weights.begin();
        for(it1 ; it1 != weights.end() ; it1++){
            pair<char,char> node = it1->first; weight = it1->second;
            G[charToInt[node.first]].push_back({charToInt[node.second], weight});
        }
        cin >> from >> to;
        young = charToInt[from]; elder = charToInt[to];


        vector<vector<int>> distsYoung = floydWarshall(G, types, 'Y', intToChar);
        vector<vector<int>> distsElder = floydWarshall(G, types, 'M', intToChar);
        
        

        ans = 1000;
        vector<char> answers;
        for(i = 0 ; i < n ; i++){
            char place = intToChar[i];
            if(types.find({from, place}) != types.end() and types.find({to, place}) != types.end() and distsYoung[young][i]+distsElder[elder][i] == ans)
                answers.push_back(place);
            else if(types.find({from, place}) != types.end() and types.find({to, place}) != types.end() and distsYoung[young][i]+distsElder[elder][i] < ans){
                ans = distsYoung[young][i]+distsElder[elder][i];
                answers.clear();
                answers.push_back(place);
            }
        }
        if(ans == 1000) cout << "You will never meet." << endl;
        else{
            cout << ans;
            for(i = 0 ; i < answers.size() ; i++) cout << " " << answers[i];
            cout << endl;
        }
        
    }
    return 0;
}

File: 10171.py
from sys import stdin

def floydWarshall(dist):
    n = len(dist)
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]


def main():
    edgesCnt = int(stdin.readline())
    while edgesCnt != 0:
        nodes, inputs = set(), list()
        for i in range(edgesCnt):
            inp = stdin.readline().strip().split()
            nodes.add(inp[2]); nodes.add(inp[3])
            inputs.append(inp)

        s, t = stdin.readline().strip().split(); nodes.add(s); nodes.add(t)
        n, charToInt, intToChar, i = len(nodes), dict(), dict(), 0
        vertices = list()
        for v in nodes: vertices.append(v)
        vertices.sort()
        for v in vertices:
            charToInt[v], intToChar[i] = i, v
            i += 1
        distYoung, distElder = [[1000 for _ in range(n)] for _ in range(n)],  [[1000 for _ in range(n)] for _ in range(n)]
        for option, direct, a, b, weight in inputs:
            u, v = charToInt[a], charToInt[b]
            distYoung[u][u], distYoung[v][v], distElder[u][u], distElder[v][v] = 0, 0, 0, 0
            if option == "Y":
                distYoung[u][v] = int(weight)
                if direct == "B": distYoung[v][u] = int(weight)
            else:
                distElder[u][v] = int(weight)
                if direct == "B": distElder[v][u] = int(weight)

        for i in range(n):
            distElder[i][i] = 0
            distYoung[i][i] = 0
        i, j = charToInt[s], charToInt[t]
        
        
        
        floydWarshall(distElder); floydWarshall(distYoung)
        ans, answers = 1000, []
        for k in range(n):
            if distYoung[i][k]+distElder[j][k] < ans:
                ans = distYoung[i][k]+distElder[j][k]
                answers = [intToChar[k]]
            elif distYoung[i][k]+distElder[j][k] == ans: answers.append(intToChar[k])
        answers.sort()
        if ans != 1000:
            print(ans, end='')
            ans = ""
            for v in answers: ans+= (" "+v)
            print(ans)
        else: print("You will never meet.")
            
        edgesCnt = int(stdin.readline())

main()